File: src/main/java/top/zbeboy/isy/service/system/MobileServiceImpl.java
package top.zbeboy.isy.service.system;


import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.CharEncoding;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;
import top.zbeboy.isy.config.ISYProperties;
import top.zbeboy.isy.domain.tables.pojos.SystemSms;
import top.zbeboy.isy.service.util.UUIDUtils;

import javax.annotation.Resource;
import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLEncoder;
import java.sql.Timestamp;
import java.time.Clock;

/**
 * Created by lenovo on 2016-05-17.
 */
@Slf4j
@Service("mobileService")
public class MobileServiceImpl implements MobileService {

    @Autowired
    private ISYProperties isyProperties;

    @Resource
    private SystemSmsService systemSmsService;

    @Async
    @Override
    public void sendShortMessage(String mobile, String content) {
        String result = "";
        try {
            String httpUrl = "http://apis.baidu.com/kingtto_media/106sms/106sms";
            String sendContent = URLEncoder.encode(content, CharEncoding.UTF_8);
            log.debug(" mobile sendContent : {}", sendContent);
            String httpArg = "mobile=" + mobile + "&content=" + sendContent;
            BufferedReader reader;
            StringBuilder sbf = new StringBuilder();
            httpUrl = httpUrl + "?" + httpArg;
            URL url = new URL(httpUrl);
            HttpURLConnection connection = (HttpURLConnection) url
                    .openConnection();
            connection.setRequestMethod("GET");
            // 填入apikey到HTTP header
            connection.setRequestProperty("apikey", isyProperties.getMobile().getApikey());
            connection.connect();
            InputStream is = connection.getInputStream();
            reader = new BufferedReader(new InputStreamReader(is, CharEncoding.UTF_8));
            String strRead;
            while ((strRead = reader.readLine()) != null) {
                sbf.append(strRead);
                sbf.append("\r\n");
            }
            reader.close();
            result = sbf.toString();
        } catch (Exception e) {
            log.info("Send sms to mobile {} is exception : {}", mobile, e);
            result = e.getMessage();
        }
        SystemSms systemSms = new SystemSms(UUIDUtils.getUUID(), new Timestamp(Clock.systemDefaultZone().millis()), mobile, result);
        systemSmsService.save(systemSms);
    }

    @Async
    @Override
    public void sendValidMobileShortMessage(String mobile, String verificationCode) {
        log.debug(" mobile valid : {} : {}", mobile, verificationCode);
        if (isyProperties.getMobile().isOpen()) {
            String content = "【ISY信息平台】 您的验证码:" + verificationCode;
            sendShortMessage(mobile, content);
        } else {
            log.debug(" 管理员已关闭短信发送 ");
        }
    }
}

File: src/main/java/top/zbeboy/isy/service/system/MobileServiceImpl.kt
package top.zbeboy.isy.service.system

import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Service
import top.zbeboy.isy.config.ISYProperties
import top.zbeboy.isy.elastic.pojo.SystemSmsElastic
import top.zbeboy.isy.glue.system.SystemSmsGlue
import top.zbeboy.isy.service.util.UUIDUtils
import java.io.*
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.sql.Timestamp
import java.time.Clock
import javax.annotation.Resource

/**
 * Created by zbeboy 2017-11-11 .
 **/
@Service("mobileService")
open class MobileServiceImpl : MobileService {

    private val log = LoggerFactory.getLogger(MobileServiceImpl::class.java)

    @Autowired
    open lateinit var isyProperties: ISYProperties

    @Resource
    open lateinit var systemSmsGlue: SystemSmsGlue

    @Async
    override fun sendShortMessage(mobile: String, content: String, action: String?,
                                  sendType: String?, codingType: String?, backEncodeType: String?) {
        var result: String?
        try {
            var ct = codingType
            var bet = backEncodeType
            if (ct == null || ct == "") {
                ct = Charsets.UTF_8.displayName()
            }
            if (bet == null || bet == "") {
                bet = Charsets.UTF_8.displayName()
            }
            val send = StringBuffer()
            if (action != null && action != "") {
                send.append("action=").append(action)
            } else {
                send.append("action=send")
            }

            send.append("&userid=").append(isyProperties.getMobile().userId)
            send.append("&account=").append(
                    URLEncoder.encode(isyProperties.getMobile().account, ct))
            send.append("&password=").append(
                    URLEncoder.encode(isyProperties.getMobile().password, ct))
            send.append("&mobile=").append(mobile)
            send.append("&content=").append(
                    URLEncoder.encode(content, ct))
            if (sendType != null && sendType.toLowerCase() == "get") {
                result = SmsClientAccessTool.getInstance().doAccessHTTPGet(
                        isyProperties.getMobile().url + "?" + send.toString(), bet)
            } else {
                result = SmsClientAccessTool.getInstance().doAccessHTTPPost(isyProperties.getMobile().url!!,
                        send.toString(), bet)
            }
        } catch (e: Exception) {
            log.error("发送短信异常:{}", e)
            result = e.message
        }

        val systemSms = SystemSmsElastic(UUIDUtils.getUUID(), Timestamp(Clock.systemDefaultZone().millis()), mobile, result)
        systemSmsGlue.save(systemSms)
    }

    @Async
    override fun sendValidMobileShortMessage(mobile: String, verificationCode: String) {
        log.debug(" mobile valid : {} : {}", mobile, verificationCode)
        if (isyProperties.getMobile().isOpen) {
            val content = "【" + isyProperties.getMobile().sign + "】 您的验证码是:" + verificationCode + "，感谢您的使用！"
            sendShortMessage(mobile, content, "", "", "", "")
        } else {
            log.debug(" 管理员已关闭短信发送 ")
        }
    }


    class SmsClientAccessTool {

        private val log = LoggerFactory.getLogger(SmsClientAccessTool::class.java)

        companion object {
            @JvmField
            var smsClientToolInstance: SmsClientAccessTool? = null

            /**
             * 采用单列方式来访问操作
             *
             * @return
             */
            @JvmStatic
            @Synchronized
            fun getInstance(): SmsClientAccessTool {

                if (smsClientToolInstance == null) {
                    smsClientToolInstance = SmsClientAccessTool()
                }
                return smsClientToolInstance as SmsClientAccessTool
            }
        }

        /**
         * POST方法
         *
         * @param sendUrl
         * ：访问URL
         * @param sendParam
         * ：参数串
         * @param backEncodeType
         * ：返回的编码
         * @return
         */
        fun doAccessHTTPPost(sendUrl: String, sendParam: String,
                             backEncodeType: String?): String {
            var encodeType = backEncodeType

            val receive = StringBuffer()
            val wr: BufferedWriter? = null
            try {
                if (encodeType == null || encodeType == "") {
                    encodeType = "UTF-8"
                }

                val url = URL(sendUrl)
                val URLConn = url
                        .openConnection() as HttpURLConnection

                URLConn.doOutput = true
                URLConn.doInput = true
                URLConn.requestMethod = "POST"
                URLConn.useCaches = false
                URLConn.allowUserInteraction = true
                HttpURLConnection.setFollowRedirects(true)
                URLConn.instanceFollowRedirects = true

                URLConn.setRequestProperty("Content-Type",
                        "application/x-www-form-urlencoded;charset=UTF-8")
                URLConn.setRequestProperty("Content-Length", sendParam.toByteArray().size.toString())

                val dos = DataOutputStream(URLConn
                        .outputStream)
                dos.writeBytes(sendParam)

                val rd = BufferedReader(InputStreamReader(
                        URLConn.inputStream, encodeType))
                var line: String? = rd.readLine()
                while (line != null) {
                    receive.append(line).append("\r\n")
                    line = rd.readLine()
                }
                rd.close()
            } catch (e: java.io.IOException) {
                receive.append("访问产生了异常-->").append(e.message)
                log.error("Send sms error . {}", e)
            } finally {
                if (wr != null) {
                    try {
                        wr.close()
                    } catch (ex: IOException) {
                        log.error("Send sms error . {}", ex)
                    }
                }
            }

            return receive.toString()
        }

        fun doAccessHTTPGet(sendUrl: String, backEncodeType: String?): String {
            var encodeType = backEncodeType

            val receive = StringBuffer()
            var `in`: BufferedReader? = null
            try {
                if (encodeType == null || encodeType == "") {
                    encodeType = "UTF-8"
                }

                val url = URL(sendUrl)
                val URLConn = url
                        .openConnection() as HttpURLConnection

                URLConn.doInput = true
                URLConn.doOutput = true
                URLConn.connect()
                URLConn.outputStream.flush()
                `in` = BufferedReader(InputStreamReader(URLConn
                        .inputStream, encodeType))

                var line: String? = `in`.readLine()
                while (line != null) {
                    receive.append(line).append("\r\n")
                    line = `in`.readLine()
                }

            } catch (e: IOException) {
                receive.append("访问产生了异常-->").append(e.message)
                log.error("Send sms error . {}", e)
            } finally {
                if (`in` != null) {
                    try {
                        `in`.close()
                    } catch (ex: java.io.IOException) {
                        log.error("Send sms error . {}", ex)
                    }
                }
            }

            return receive.toString()
        }
    }


}